raster-foundry/raster-foundry
# Notify users in intercom when their annotation projects are ready

## Overview

This PR starts the work to notify users in intercom when their projects are ready.

### Checklist


### Demo

![notification](https://user-images.githubusercontent.com/5702984/76540092-aa440580-644f-11ea-9378-4e14e56de3c1.gif)

### Notes

~It's in Scala because _my god_ that's a complicated datamodel. I'll show a response. It's _huge_. I don't want to remember over time what values are present in this thing in python so instead I'm making the values I'm grabbing from the response json super explicit and we can add more as necessary.~ It turns out that the endpoint we're interacting with is a lot less complicated, but I didn't rewrite it in python after learning that. This is reusable in other batch jobs now, which has value as well. Also, hilariously, there's [no python sdk](https://developers.intercom.com/building-apps/docs/sdks-plugins) anyway.

~This might also move out of `batch` someday if we need it elsewhere but for now I'm thinking it jumps in at the end of the `CreateTaskGrid` job.~

~Still WIP until I make a job that sends notifications that the python code in app-tasks can invoke, but testable at least~

## Testing Instructions

- set up your groundwork to point to local rf (I can send you @Lknechtli's .env.local file that includes commentable regions for every environment if you don't have it)
- grab the new `.env` with `AWS_PROFILE=raster-foundry RF_SETTINGS_BUCKET=<settings bucket> ./scripts/bootstrap --env`
- assemble batch and anything else you haven't assembled recently
- build the batch container
- bring up your api server
- bring up your rf frontend
- log in to rf as the dev user
- log in to Groundwork locally as the dev user
- create a project in Groundwork
- grab the upload id for your new project from `./scripts/psql` -- `select id from uploads order by created_at desc limit 1;`
- process your upload: `./scripts/console batch 'rf process-upload <your upload id>'`
- observe that you got a notification that it's done (ignore the bad url -- configured rf host is odd in development because the container doesn't know how to get to the api under localhost, but it's correct in staging and prod)
- throw an exception into the try block in the python upload processing script -- add a `raise Exception("lol")` [here](https://github.com/raster-foundry/raster-foundry/blob/ab2bf86a8e7644fddbfa319c890513e4360c2aa0/app-tasks/rf/src/rf/commands/process_upload.py#L55)
- rebuild the batch container
- process your upload again but this time saying it's the third try -- `./scripts/console batch 'AWS_BATCH_JOB_ATTEMPT=3  rf process-upload <your upload id>'`
- observe that you get a notification that processing failed

Closes raster-foundry/annotate#640
